## The scripts in this folder:
In this folder, the scripts are provided with which different genomic variations in transformation hybrids can be detected.  
First, the masterlist for the recipient reference with respect to the donor is detected.  
Next, different kinds of genomic variations are detected from the alignments of hybrid sequencing reads to the reference genomes.
<br />
<kbd>
<img src="https://github.com/Easybel/DetectionGV/blob/main/1_Detection/Detection_Recombination.png" width="900">
</kbd>
All scripts are implemented in MATLAB. The order of the scripts indicates the order in which scripts are usually run.

[**A0_VariantFiltering.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0_VariantFiltering.m)
- This script takes the list of called variants **_bcftools.vcf** and filters the entries. Indels can be excluded and the filter acts upon phred SNP quality scores, read depth and the ambiguity of SNPs (exclude SNPs that make up a too small amount of all aligned reads at a position). Additionally, the Look-Ahead filter can be used, that only passes SNPs that are still present in the same sample sequenced at a later timepoint. 

[**A0b_MasterListFiltering.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0b_MasterListFiltering.m)
- The script is very similar to 
[**A0_VariantFiltering.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0_VariantFiltering.m) and is run on the donor sequencing reads aligned to the recipient used for the hybrid organisms. It is expected to be run with the same filter paramters as used for the variant filtering of the hybrid strains. It gives a special masterlist **.mat** output that can be used for the [**A1_SNP2CNP.m** ](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A1_SNP2CNP.m) script. 

[**A0c_AccessoryGenome.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0c_AccessoryGenome.m)
- Here, we detect the accessory genome parts of the recipient reference with respect to the donr (or vice versa). The accessory regions found here are excluded from the replacement detection performed with [**A1_SNP2CNP.m** ](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A1_SNP2CNP.m). As an input, this script takes the **.vcf** output from the **mpileup** function run in [2_CombiScriptSamtools.sh](https://github.com/Easybel/DetectionGV/blob/main/0_WGSPipeline/2_CombiScriptSamtools.sh). For all positions that have a read depth below the threshold set in [**A0_VariantFiltering.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0_VariantFiltering.m), the script identifies consecutive regions longer than 150 bp. These are collected as accessory regions.

[**A0d_Multimapper.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A0d_Multimapper.m)
- This script allows us to find the regions on a genome to which multimapping reads align.

[**A1_SNP2CNP.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/A1_SNP2CNP.m)
- With the filtered variant list and the masterlist, we detect the replacements of donor segments into the recipient. Here, we exclude accessory genome and multimapping regions.

[**A2_Lists2Genes.m**](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/**A2_Lists2Genes.m**)
- This scripts allows us to find the genes that are detected as genetically modified. The script accepts as input a plain list of start and end positions, the output generated by **A1_SNP2CNP.m**, as well as a list of SNPs. 
 
[**A3_Cov2DelDup.m** ](https://github.com/Easybel/DetectionGV/blob/main/1_Detection/**A3_Cov2DelDup.m**)
- This script contains an algorithm that detects deletions and duplications. The input is the coverage of the aligned reads to the reference.

---------------------------------------------------
